// @pm2/io/build/main/src/actions/eventLoopInspector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const debug_1 = require("debug");
const debug = debug_1.default('axm:eventloopaction');
const module_1 = require("../utils/module");
class Inspector {
    constructor(actionFeature) {
        this.MODULE_NAME = 'event-loop-inspector';
        this.actionFeature = actionFeature;
    }
    init() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                module_1.default.detectModule(this.MODULE_NAME, (err, inspectorPath) => {
                    if (err) {
                        debug(err);
                        return reject(err);
                    }
                    const res = this.exposeActions(inspectorPath);
                    if (res instanceof Error) {
                        return reject(res);
                    }
                    return resolve();
                });
            }).catch((e) => console.error(e));
        });
    }
    exposeActions(inspectorPath) {
        let inspector = module_1.default.loadModule(inspectorPath, this.MODULE_NAME, [true]);
        if (inspector instanceof Error || !inspector) {
            return inspector;
        }
        /**
         * Heap snapshot
         */
        return this.actionFeature.action('km:event-loop-dump', function (reply) {
            const dump = inspector.dump();
            return reply({
                success: true,
                dump: dump
            });
        });
    }
}
exports.default = Inspector;
